File: packages/tool-sdk/src/python/ideamine_tools/telemetry.py
"""
IdeaMine Tools SDK - Python Telemetry
OpenTelemetry integration for distributed tracing
"""

import os
from typing import Any, Dict, Optional
from contextlib import contextmanager

from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.semconv.resource import ResourceAttributes
from opentelemetry.trace import Span, Status, StatusCode, SpanKind
import structlog


class TelemetryManager:
    """OpenTelemetry manager for tool tracing"""

    def __init__(
        self,
        enabled: bool,
        service_name: str,
        endpoint: Optional[str] = None,
        logger: Optional[structlog.BoundLogger] = None
    ):
        self.enabled = enabled
        self.logger = logger or structlog.get_logger()
        self.tracer: Optional[trace.Tracer] = None
        self.provider: Optional[TracerProvider] = None

        if self.enabled:
            try:
                # Create resource with service information
                resource = Resource.create({
                    ResourceAttributes.SERVICE_NAME: service_name,
                    ResourceAttributes.SERVICE_VERSION: "1.0.0",
                })

                # Initialize provider
                self.provider = TracerProvider(resource=resource)

                # Configure exporter
                otlp_endpoint = (
                    endpoint or
                    os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT") or
                    "http://localhost:4318/v1/traces"
                )
                exporter = OTLPSpanExporter(endpoint=otlp_endpoint)

                # Add batch processor
                self.provider.add_span_processor(BatchSpanProcessor(exporter))

                # Register provider
                trace.set_tracer_provider(self.provider)

                # Get tracer
                self.tracer = trace.get_tracer(service_name)

                self.logger.info(
                    "OpenTelemetry initialized",
                    service=service_name,
                    endpoint=otlp_endpoint
                )
            except Exception as e:
                self.logger.error("Failed to initialize OpenTelemetry", error=str(e))
                self.enabled = False

    def start_execution_span(
        self,
        tool_name: str,
        version: str,
        run_id: str,
        attributes: Optional[Dict[str, Any]] = None
    ) -> Optional[Span]:
        """Start a new span for tool execution"""
        if not self.enabled or not self.tracer:
            return None

        try:
            span = self.tracer.start_span(
                f"tool.execute.{tool_name}",
                kind=SpanKind.INTERNAL,
                attributes={
                    "tool.name": tool_name,
                    "tool.version": version,
                    "run.id": run_id,
                    **(attributes or {})
                }
            )
            return span
        except Exception as e:
            self.logger.error("Failed to start span", error=str(e))
            return None

    def start_validation_span(
        self,
        schema_type: str,
        tool_name: str
    ) -> Optional[Span]:
        """Start a span for validation"""
        if not self.enabled or not self.tracer:
            return None

        try:
            span = self.tracer.start_span(
                f"tool.validate.{schema_type}",
                kind=SpanKind.INTERNAL,
                attributes={
                    "validation.type": schema_type,
                    "tool.name": tool_name,
                }
            )
            return span
        except Exception as e:
            self.logger.error("Failed to start validation span", error=str(e))
            return None

    def end_span(self, span: Optional[Span], attributes: Optional[Dict[str, Any]] = None) -> None:
        """End span with success"""
        if not span:
            return

        try:
            if attributes:
                span.set_attributes(attributes)
            span.set_status(Status(StatusCode.OK))
            span.end()
        except Exception as e:
            self.logger.error("Failed to end span", error=str(e))

    def end_span_with_error(self, span: Optional[Span], error: Exception) -> None:
        """End span with error"""
        if not span:
            return

        try:
            span.set_status(
                Status(StatusCode.ERROR, description=str(error))
            )
            span.record_exception(error)
            span.end()
        except Exception as e:
            self.logger.error("Failed to end span with error", error=str(e))

    def add_event(
        self,
        span: Optional[Span],
        name: str,
        attributes: Optional[Dict[str, Any]] = None
    ) -> None:
        """Add event to current span"""
        if not span:
            return

        try:
            span.add_event(name, attributes=attributes or {})
        except Exception as e:
            self.logger.error("Failed to add event to span", error=str(e))

    def shutdown(self) -> None:
        """Shutdown telemetry gracefully"""
        if self.provider:
            try:
                self.provider.shutdown()
                self.logger.info("OpenTelemetry shutdown complete")
            except Exception as e:
                self.logger.error("Failed to shutdown OpenTelemetry", error=str(e))

    @contextmanager
    def trace_execution(
        self,
        tool_name: str,
        version: str,
        run_id: str,
        attributes: Optional[Dict[str, Any]] = None
    ):
        """Context manager for automatic span tracking"""
        span = self.start_execution_span(tool_name, version, run_id, attributes)
        try:
            yield span
            self.end_span(span)
        except Exception as e:
            self.end_span_with_error(span, e)
            raise

File: packages/tool-sdk/src/ts/telemetry.ts
/**
 * IdeaMine Tools SDK - OpenTelemetry Integration
 * Distributed tracing for tool executions
 */

import { trace, context, Span, SpanStatusCode, SpanKind } from '@opentelemetry/api';
import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { Resource } from '@opentelemetry/resources';
import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions';
import { BatchSpanProcessor } from '@opentelemetry/sdk-trace-node';
import { Logger } from './types';

export interface TelemetryConfig {
  enabled: boolean;
  serviceName: string;
  endpoint?: string;
  sampleRate?: number;
}

export class TelemetryManager {
  private provider?: NodeTracerProvider;
  private tracer: any;
  private enabled: boolean;
  private logger: Logger;

  constructor(config: TelemetryConfig, logger: Logger) {
    this.enabled = config.enabled;
    this.logger = logger;

    if (this.enabled) {
      try {
        // Create resource with service information
        const resource = Resource.default().merge(
          new Resource({
            [SemanticResourceAttributes.SERVICE_NAME]: config.serviceName,
            [SemanticResourceAttributes.SERVICE_VERSION]: '1.0.0',
          })
        );

        // Initialize provider
        this.provider = new NodeTracerProvider({
          resource,
        });

        // Configure exporter
        const endpoint = config.endpoint || process.env.OTEL_EXPORTER_OTLP_ENDPOINT || 'http://localhost:4318/v1/traces';
        const exporter = new OTLPTraceExporter({
          url: endpoint,
        });

        // Add batch processor
        this.provider.addSpanProcessor(new BatchSpanProcessor(exporter));

        // Register provider
        this.provider.register();

        // Get tracer
        this.tracer = trace.getTracer(config.serviceName);

        this.logger.info('OpenTelemetry initialized', {
          service: config.serviceName,
          endpoint,
        });
      } catch (error) {
        this.logger.error('Failed to initialize OpenTelemetry', { error });
        this.enabled = false;
      }
    }
  }

  /**
   * Start a new span for a tool execution
   */
  startExecutionSpan(
    toolName: string,
    version: string,
    runId: string,
    attributes?: Record<string, any>
  ): Span | null {
    if (!this.enabled || !this.tracer) {
      return null;
    }

    try {
      const span = this.tracer.startSpan(
        `tool.execute.${toolName}`,
        {
          kind: SpanKind.INTERNAL,
          attributes: {
            'tool.name': toolName,
            'tool.version': version,
            'run.id': runId,
            ...attributes,
          },
        }
      );

      return span;
    } catch (error) {
      this.logger.error('Failed to start span', { error });
      return null;
    }
  }

  /**
   * Start a span for validation
   */
  startValidationSpan(
    schemaType: 'input' | 'output',
    toolName: string
  ): Span | null {
    if (!this.enabled || !this.tracer) {
      return null;
    }

    try {
      const span = this.tracer.startSpan(
        `tool.validate.${schemaType}`,
        {
          kind: SpanKind.INTERNAL,
          attributes: {
            'validation.type': schemaType,
            'tool.name': toolName,
          },
        }
      );

      return span;
    } catch (error) {
      this.logger.error('Failed to start validation span', { error });
      return null;
    }
  }

  /**
   * End span with success
   */
  endSpan(span: Span | null, attributes?: Record<string, any>): void {
    if (!span) return;

    try {
      if (attributes) {
        span.setAttributes(attributes);
      }
      span.setStatus({ code: SpanStatusCode.OK });
      span.end();
    } catch (error) {
      this.logger.error('Failed to end span', { error });
    }
  }

  /**
   * End span with error
   */
  endSpanWithError(span: Span | null, error: Error | string): void {
    if (!span) return;

    try {
      const errorMessage = typeof error === 'string' ? error : error.message;
      const errorStack = typeof error === 'string' ? undefined : error.stack;

      span.setStatus({
        code: SpanStatusCode.ERROR,
        message: errorMessage,
      });

      span.recordException({
        name: 'ToolError',
        message: errorMessage,
        stack: errorStack,
      } as any);

      span.end();
    } catch (err) {
      this.logger.error('Failed to end span with error', { err });
    }
  }

  /**
   * Add event to current span
   */
  addEvent(span: Span | null, name: string, attributes?: Record<string, any>): void {
    if (!span) return;

    try {
      span.addEvent(name, attributes);
    } catch (error) {
      this.logger.error('Failed to add event to span', { error });
    }
  }

  /**
   * Shutdown telemetry gracefully
   */
  async shutdown(): Promise<void> {
    if (this.provider) {
      try {
        await this.provider.shutdown();
        this.logger.info('OpenTelemetry shutdown complete');
      } catch (error) {
        this.logger.error('Failed to shutdown OpenTelemetry', { error });
      }
    }
  }
}

/**
 * Helper to run function with automatic span tracking
 */
export async function withSpan<T>(
  telemetry: TelemetryManager,
  spanName: string,
  attributes: Record<string, any>,
  fn: (span: Span | null) => Promise<T>
): Promise<T> {
  const span = telemetry.startExecutionSpan(spanName, '1.0.0', 'unknown', attributes);

  try {
    const result = await fn(span);
    telemetry.endSpan(span);
    return result;
  } catch (error) {
    telemetry.endSpanWithError(span, error as Error);
    throw error;
  }
}
